File: backend/course.py
# Course object

class Course:
    def __init__(self, subject=None,  class_credits=None, crn=None, days=None,
                 start_time=None,     end_time=None,      lab_days=None,
                 lab_start_time=None, lab_end_time=None,  instructor=None,
                 room=None,           lab_room=None,      addl_fees=None, 
                 cap=None, enrl=None, avail=None,         waitlist=None,
                 restrictions=None,   attributes=None,    prerequisites=None):
        self.subject = subject
        self.class_credits = class_credits
        self.crn = crn
        self.days = days
        self.start_time = start_time
        self.end_time = end_time
        self.lab_days = lab_days
        self.lab_start_time = lab_start_time
        self.lab_end_time = lab_end_time
        self.instructor = instructor
        self.room = room
        self.lab_room = lab_room
        self.addl_fees = addl_fees
        self.cap = cap
        self.enrl = enrl
        self.avail = avail
        self.waitlist = waitlist
        self.restrictions = restrictions
        self.attributes = attributes
        self.prerequisites = prerequisites

    def __repr__(self):
        return (
            f"Course(subject={self.subject}, class_credits={self.class_credits}, "
            f"crn={self.crn}, days={self.days}, start_time={self.start_time}, "
            f"end_time={self.end_time}, lab_days={self.lab_days}, "
            f"lab_start_time={self.lab_start_time}, lab_end_time={self.lab_end_time}, "
            f"instructor={self.instructor}, room={self.room}, addl_fees={self.addl_fees}, "
            f"cap={self.cap}, enrl={self.enrl}, avail={self.avail}, "
            f"waitlist={self.waitlist}, restrictions={self.restrictions}, "
            f"attributes={self.attributes}, prerequisites={self.prerequisites})"
        )

    def to_dict(self):
        return {
            "subject": self.subject,
            "class_credits": self.class_credits,
            "crn": self.crn,
            "days": self.days,
            "start_time": self.start_time,
            "end_time": self.end_time,
            "lab_days": self.lab_days,
            "lab_start_time": self.lab_start_time,
            "lab_end_time": self.lab_end_time,
            "instructor": self.instructor,
            "room": self.room,
            "lab_room": self.lab_room,
            "addl_fees": self.addl_fees,
            "cap": self.cap,
            "enrl": self.enrl,
            "avail": self.avail,
            "waitlist": self.waitlist,
            "restrictions": self.restrictions,
            "attributes": self.attributes,
            "prerequisites": self.prerequisites
        }

    def conflicts(self, other: 'Course') -> bool:
        # Two courses confilict if they have the same subject
        if self.subject == other.subject:
            return True
        # Compare self days with the others days
        for day in self.days:
            if day in other.days:
                if (((self.start_time >= other.start_time) and 
                     (self.start_time <= other.end_time)) or
                     ((self.end_time >= other.start_time) and 
                      (self.end_time <= other.end_time))):
                    return True
        # Compare self's lab with other's days
        if self.lab_days:
            for day in self.lab_days:
                if day in other.days:
                    if (((self.lab_start_time >= other.start_time) and 
                         (self.lab_start_time <= other.end_time)) or
                        ((self.lab_end_time >= other.start_time) and 
                         (self.lab_end_time <= other.end_time))):
                        return True
        # Compare other's lab with self's days
        if other.lab_days:
            for day in other.lab_days:
                if day in self.days:
                    if (((self.start_time >= other.lab_start_time) and 
                         (self.start_time <= other.lab_end_time)) or
                        ((self.end_time >= other.lab_start_time) and 
                         (self.end_time <= other.lab_end_time))):
                        return True
        # Compare lab times between both labs
        if self.lab_days and other.lab_days:
            for day in self.lab_days:
                if day in other.lab_days:
                    if (((self.lab_start_time >= other.lab_start_time) and 
                         (self.lab_start_time <= other.lab_end_time)) or
                        ((self.lab_end_time >= other.lab_start_time) and 
                         (self.lab_end_time <= other.lab_end_time))):
                        return True
        return False

File: go-backend/internal/models/course.go
// Package models for re-usable objects
package models

import "strings"

// Holds one session worth of data (repeatable time range over days)
type Session struct {
	Days       string // Days of the week in format (e.g. "MTWRF")
	StartTime  int    // Start time in 24-hour format (0000-2359)
	EndTime    int    // End time in 24-hour format (0000-2359)
	Instructor string // Session Instructor
	Location   string // Room or location
	IsAsync    bool   // True if the course is asynchronous
	IsTimeTBD  bool   // True if times or days are tbd
}

// Holds all of a single courses data
type Course struct {
	Subject        string    // Course Subject
	Title          string    // Course Title
	Credits        int       // Number of credits
	CRN            int       // Course Record Number
	Sessions       []Session // All sessions for this course (nil if asnyc)
	GPA            float64   // Average GPA for this course
	Capacity       int       // Maximum enrollment capacity
	Enrolled       int       // Number of currently enrolled students
	AvailableSeats int       // Number of seats available
	WaitlistCount  int       // Number of students on waitlist
	AdditionalFees string    // Any noted additional fees
	Restrictions   string    // Any noted restrictions
	Attributes     string    // Any noted attributes
	Prerequisites  string    // Any noted prerequisites
}

// Initialize fully empty course
func NewCourse() *Course {
	return &Course{
		Subject:        "",
		Title:          "",
		Credits:        0,
		CRN:            0,
		Sessions:       nil,
		GPA:            0.0,
		Capacity:       0,
		Enrolled:       0,
		AvailableSeats: 0,
		WaitlistCount:  0,
		AdditionalFees: "",
		Restrictions:   "",
		Attributes:     "",
		Prerequisites:  "",
	}
}

// Returns whether the courses conflict with one another
func (c Course) Conflicts(other Course) bool {
	// Conflicts on the same subject
	if c.Subject == other.Subject {
		return true
	}

	for _, s1 := range c.Sessions {
		for _, s2 := range other.Sessions {
			if sessionsConflict(s1, s2) {
				return true
			}
		}
	}

	return false
}

// sessionsConflict checks if two sessions conflict
func sessionsConflict(s1, s2 Session) bool {
	// Two courses can't be known to conflict if they are online or tbd
	if s1.IsAsync || s1.IsTimeTBD || s2.IsAsync || s2.IsTimeTBD {
		return true
	}

	// Check if s2 contains the day in s1. If so, checks time overlap
	for _, day1 := range s1.Days {
		if strings.ContainsRune(s2.Days, day1) {
			if timesConflict(s1.StartTime, s1.EndTime, s2.StartTime, s2.EndTime) {
				return true
			}
		}
	}
	return false
}

// timesConflict checks if two time ranges overlap
func timesConflict(start1, end1, start2, end2 int) bool {
	return (((start1 >= start2) && (start1 <= end2)) || ((end1 >= start2) && (end1 <= end2)))
}
